grudge/models/__init__.py

```diff
@@ -49,13 +49,22 @@ def max_characteristic_velocity(self, t, fields, dcoll):
 
     def estimate_rk4_timestep(self, dcoll, t=None, fields=None):
         """Estimate the largest stable timestep for an RK4 method."""
-        from grudge.dt_utils import (dt_non_geometric_factor,
+        from grudge.dt_utils import (dt_non_geometric_factors,
                                      dt_geometric_factors)
+        from meshmode.dof_array import DOFArray
         import grudge.op as op
 
+        actx = dcoll._setup_context
         max_lambda = self.max_characteristic_velocity(t, fields, dcoll)
-        dt_factor = \
-            (dt_non_geometric_factor(dcoll)
-             * op.nodal_min(dcoll, "vol", dt_geometric_factors(dcoll)))
+        # Scale each group array of geometric factors by the corresponding
+        # group non-geometric factor
+        dt_factors = DOFArray(
+            actx,
+            data=(
+                cng * geo_facts
+                for cng, geo_facts in zip(dt_non_geometric_factors(dcoll),
+                                          dt_geometric_factors(dcoll))
+            )
+        )
 
-        return dt_factor * (1 / max_lambda)
+        return op.nodal_min(dcoll, "vol", dt_factors * (1 / max_lambda))
```

Comment:
For fluids, I feel like the multiplication by `1/max_lambda` should probably also be inside the `nodal_min`.

Reply:
It is :)